### PR TITLE
Remote Government

### LINKED ISSUE #90: Remove or replace sudo with multisig after supporting remote governance.

### PR DESCRIPTION
**Pull request by acgxv**
Close #90.

### CODE REVIEW

File: runtime/common/src/remote_government/mod.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+			Ok(Pays::No.into())
+		}
+	}
+	impl<T: Config> Pallet<T> {
+		fn check_sync() -> bool {
+			<PreviousBridgeFinalized<T>>::try_mutate(|previous_bridge_finalized| {
+				let best_finalized = T::BridgeFinalized::get();
+
+				if *previous_bridge_finalized == best_finalized {
+					Err(())
```
Comment (hackfisher) [inline runtime/common/src/remote_government/mod.rs:194]: Why return error here?

**Review by hackfisher — CHANGES_REQUESTED** [review]
As I mentioned, I'm also worried about other bugs in message protocol, not only the grandpa stalled possibility.

Request to remove the Emergency and check sync.

File: runtime/common/src/remote_government/mod.rs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+			Self::deposit_event(Event::RemoteCallEnacted {
+				result: res.map(|_| ()).map_err(|e| e.error),
+			});
+
+			// Sudo user does not pay a fee.
+			Ok(Pays::No.into())
+		}
+	}
+	impl<T: Config> Pallet<T> {
+		fn check_sync() -> bool {
```
Comment (hackfisher) [inline runtime/common/src/remote_government/mod.rs]:
```suggestion
		fn try_sync_new_bridge_grandpa_header() -> bool {
```

File: runtime/common/src/remote_government/mod.rs (hunk 1 above)
Comment (acgxv) [reply runtime/common/src/remote_government/mod.rs:194]: If they are the same, skip storage writing.

**Comment by hackfisher** [comment]
> As I mentioned, I'm also worried about other bugs in message protocol, not only the grandpa stalled possibility.
> 
> Request to remove the Emergency and check sync.

There are two fallbacks if the message protocol failed.

1. The owner of bridge-grandpa can halt the relay of headers, and trigger the emergency proactively.
2. Request help from relaychain(Kusama), to modify the runtime code from relaychain first, and do a hardfork on parachain.

If these two fallbacks can work, I think it's good to keep the emergency design, the naming and documents need to be update to detail the above cases.

File: runtime/pangolin-parachain/src/lib.rs [hunk 3]
```diff
@@ -186,7 +186,9 @@ frame_support::construct_runtime! {
 		BridgePangolinDispatch: pallet_bridge_dispatch::<Instance1>::{Pallet, Event<T>} = 22,
 
 		PangolinFeeMarket: pallet_fee_market::<Instance1>::{Pallet, Call, Storage, Event<T>} = 23,
-		FromPangolinIssuing: dc_common_runtime::helixbridge::{Pallet, Call, Storage, Event<T>} = 24,
+		FromPangolinIssuing: dp_common_runtime::helixbridge::{Pallet, Call, Storage, Event<T>} = 24,
+
+		RemoteGovernment: dp_common_runtime::remote_government::{Pallet, Call, Event<T>} = 25,
```
Comment (xiaoch05) [inline runtime/pangolin-parachain/src/lib.rs]:
```suggestion
		RemoteGovernment: dp_common_runtime::remote_government::{Pallet, Call, Event<T>} = 25,
		RemoteGovernment: dp_common_runtime::remote_government::{Pallet, Call, Storage, Event<T>} = 25,
```

File: runtime/pangolin-parachain/src/pallets/remote_government.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	pub const CheckInterval: BlockNumber = 2 * HOURS;
+}
+
+impl Config for Runtime {
+	type BridgeAccountIdConverter = bp_pangolin::AccountIdConverter;
+	type BridgeFinalized = PangolinBestFinalized;
+	type BridgedChainId = PangolinChainId;
+	type Call = Call;
+	type CheckInterval = CheckInterval;
+	type EmergencySafeguardOrigin = EnsureSpecific;
```
Comment (hackfisher) [inline runtime/pangolin-parachain/src/pallets/remote_government.rs:44]: Change to multisig or technical commitee in the runtime after next

File: runtime/pangolin-parachain/src/pallets/remote_government.rs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+pub struct PangolinBestFinalized;
+impl Get<Hash> for PangolinBestFinalized {
+	fn get() -> Hash {
+		<pallet_bridge_grandpa::BestFinalized<Runtime, WithPangolinGrandpa>>::get()
+	}
+}
+
+frame_support::parameter_types! {
+	pub const CheckInterval: BlockNumber = 2 * HOURS;
```
Comment (hackfisher) [inline runtime/pangolin-parachain/src/pallets/remote_government.rs:35]:
```suggestion
	pub const CheckInterval: BlockNumber = 6 * HOURS;
```

Comment (acgxv) [reply runtime/pangolin-parachain/src/pallets/remote_government.rs:35]: Each test will take 6 hours.
Should we change this?

Comment (hackfisher) [reply runtime/pangolin-parachain/src/pallets/remote_government.rs:35]: Ah, this is for pangolin, not Crab.

No need to change then

**Review by hackfisher — APPROVED** [review]

File: runtime/pangolin-parachain/src/pallets/remote_government.rs (hunk 4 above)
Comment (hackfisher) [reply runtime/pangolin-parachain/src/pallets/remote_government.rs:44]: https://github.com/darwinia-network/darwinia-parachain/issues/99

**Comment by hackfisher** [comment]
@AurevoirXavier Conflicts need to be resolved before merge.

**Pull request merged**